### DISCUSSION TITLE
PyInstaller with *.pyd files

### DISCUSSION THREAD

**Discussion by Mouaz-Alnouri** [pyinstaller/pyinstaller#7010] [Help]
hello, I used to use Pyinstaller with my application to produce an exe and that was working fine with me till I switched all the previous application's python modules to .pyd (using cython) because I want to secure my files.

But now after I've done that, Pyinstaller will not be able to see the imports inside those files,  So the exe will miss lots of libraries and my scripts. 

How can i make the pyinstaller bundle .pyd files and their imports? 

Please if my question isn't clear let me know.

**Comment by rokm** [comment]
> But now after I've done that, Pyinstaller will not be able to see the imports inside those files, So the exe will miss lots of libraries and my scripts.

That's directly implied by the security you wanted to achieve by compiling your modules using cython. PyInstaller cannot analyze binary extensions (cython-compiled or otherwise) for the imports that they use, so they end up missing. You'd need to generate a list of modules used by your ctyhon-compiled modules, and add those to `hiddenimports`.

We do have a helper function for indirect analysis of extensions' imports, [get_pyextension_imports](https://github.com/pyinstaller/pyinstaller/blob/c36f03d2710728ead70687689b9157d05a9b5793/PyInstaller/utils/hooks/__init__.py#L181), that works by recording `sys.modules` before and after importing the extension (and the difference are the modules that extension ended up importing) but it is probably too heavy-handed for effective use on large cythonized packages. But you could create a script that similarly imports all your cythonized modules and generates a list of modules that you can import into the spec file and pass to `hiddenimports`.

Eventually, we might implement automatic cython support as part of PyInstaller's module collection, which should be made possible by the newly introduced module collection mode flag (#6945). That would give the benefit of automatic analysis made on original file, and the module ending up collected as cythonized extension. But that's just an idea for now, so you'll have to resort to generating a hiddenimport lists by yourself...
